### PR TITLE
Adding DataReceiver with DataProviderV2

### LINKED ISSUE #196: Decoupling data schema from data format
In order to parse a JSON blob in Serde, one needs to know the data schema (struct definition).

Currently, the data provider passes Rust structs (encoded as `Any`s) in the Reponse objects through the pipeline.

Together, these two statements mean that the source data provider, the one reading the JSON blob from the file system, needs to know ahead of time the mapping from JSON files to structs.

This is undesirable because:

1. It requires some form of "key-to-struct" dispatch in the source data provider, whether a `match` statement, table lookup, etc.  This dispatch needs to be maintained and could be a source of failures or performance bottlenecks.
2. Having that dispatch means that a lot of struct parsing code needs to be carried in the source data provider, even if it is not being used downstream.  This goes against the grain of having dead code elimination also eliminate unused data structs.

I've considered a few solutions:

1. Make data providers return a blob instead of a parsed struct. Parsing of the blob would occur at the end of the chain when the type is known.
    - We'd need to either declare that the data blobs are always in one specific format (e.g. bincode), or have the source data provider also pass a deserializer function.
    - @mihnita suggested a Content-Type approach where the requestor can say that they support JSON or Bincode, and the provider needs to provide a blob in one of those two formats.
2. Send the struct definition along with the request, such that the struct can be used at the point where the data is loaded from the filesystem.
    - Essentially the opposite of option 1.
3. Use a JSON Value or some other self-describing format instead of JSON through the full stack of the data provider.
    - We lose type safety and, more importantly, data validation.
    - @kpozin pointed out that this option is basically equivalent to putting the parsing code into application logic.
4. Schema converter: pass around a blob until it needs to be parsed to a struct, and then pass around the struct (see Kubernetes apiserver)
    - Hybrid approach based on option 1 (thanks @filmil).
5. Stick with the runtime struct dispatch living in the data provider.

I think it may be useful to pass around structs, because if we get to a point where we can pre-build data into \*.rs files (#78), we'd like to pass those verbatim through the pipeline.

### PR DESCRIPTION
**Pull request by sffc**
Will Fix #196

This PR essentially implements option 2 from that issue.  Instead of DataProvider returning an opaque Any type, it takes a mutable reference argument that deserializes the data directly into the final container.

Before, where you had:

```rust
    let response = provider
        .load(&req)
        .expect("The key should be present");
    let plurals_data: &structs::plurals::PluralRuleStringsV1 = response
        .borrow_payload()
        .expect("The data should be valid");
```

Now you have:

```rust
    let mut receiver: DataReceiverImpl<structs::plurals::PluralRuleStringsV1> =
        DataReceiverImpl { payload: None };
    provider
        .load_v2(&req, &mut receiver)
        .expect("The data should be valid");
    let plurals_data: &structs::plurals::PluralRuleStringsV1 = &receiver.payload
        .expect("The data should be present");
```

I left the requirement that the struct implement Any, i.e., all data must be owned data.  I believe this condition is still required in order to allow already-deserialized data to cross the trait object boundary.  For example, DataReceiver has three methods to set the data:

```rust
pub trait DataReceiver<'d, 'de> {
    fn set_to(
        &mut self,
        deserializer: &mut dyn erased_serde::Deserializer<'de>,
    ) -> Result<(), erased_serde::Error>;

    fn set_to_any(&mut self, any: &'d dyn Any) -> Result<(), Error>;

    fn set_to_boxed(&mut self, boxed: Box<dyn Any>) -> Result<(), Error>;
}
```

I don't know how to enable DataReceiver to be a trait object but also allow it to accept a borrowed or owned object outside of the deserializer function.

### CODE REVIEW

**Comment by sffc** [comment]
In the first round of reviews, I'm looking for confirmation that you like my approach.  If you do, then I will finish this PR and replace the old DataProvider trait.

File: components/provider/src/lib.rs
Comment (Manishearth) [inline components/provider/src/lib.rs:237]: Should we instead have

```rust
fn load_v2<R: DataReceiver<'d, 'static>>(&self, req, receiver: &mut R);
fn load_v2_dyn(&self, req, receiver: &mut DataReceiver<..>);
```

with the latter written as a default implementation? This way for performance you can avoid the virtual dispatch.

File: components/provider/src/lib.rs
Comment (Manishearth) [inline components/provider/src/lib.rs:115]: Can we also split this into a generic and non-generic version? The generic `set_to` version will have to have a `where Self: Sized` so that you can still make a trait object.

This lets people write receivers that are tailored to the data being received. Serde is _really fast_ if you use generics, and that can be desired at times.

**Review by Manishearth — DISMISSED** [review]
I think this looks good, but if possible I think we should have generic versions of the `dyn`-receiving functions (they may need a `where Self: Sized` to work) so that for a fast path you can avoid the dynamic dispatch if you really want to. Serde can be blazingly fast when you don't use too much dynamic dispatch since the compiler often compiles it down to a specialized parser.

It's possible that dynamic dispatch is necessary in all the cases we will be using this API anyway, in which case we should probably pick simplicity.

Curious what @zbraniecki thinks about this.

File: components/provider/src/lib.rs
Comment (sffc) [reply components/provider/src/lib.rs:237]: Potential issues with this approach:

- Pretty sure that generics of any kind aren't allowed on functions of traits that are intended to be used as trait objects.
- I'm concerned about code size explosion, since we have many possible concrete data struct types.
- Is dynamic dispatch so bad that we should design around it?  (I haven't measured it)

File: components/provider/src/lib.rs
Comment (sffc) [reply components/provider/src/lib.rs:115]: This function uses erased_serde.  I don't have a compete mental model of how that works, but on one side of `set_to` (the concrete implementation), you have the struct deserialization code, and on the other side (the call site in the DataProvider), you have the JSON/Bincode deserialization code.  I assume they talk to each other via dynamic dispatch.  So does that mean erased_serde is slower than plain serde?  I haven't taken measurements, but I haven't read anything in particular that would suggest that is the case.

File: components/provider/src/lib.rs
Comment (Manishearth) [reply components/provider/src/lib.rs:237]: Generic functions are allowed, provided you `where Self: Sized` them.

It will only be a code size explosion if used, and I think the cost of dynamic dispatch is not _high_, but it's Rusty to give people the option. And I feel like in our code we will see some cases where we prefer that.

**Review by zbraniecki — DISMISSED** [review]
lgtm, thank you for your patience with the long review cycle!

Do you plan to keep both versions? If not, I would prefer to avoid having to already start with versioning of the APIs.

**Comment by sffc** [comment]
> Do you plan to keep both versions? If not, I would prefer to avoid having to already start with versioning of the APIs.

I plan to delete the old one after I finish migrating the rest of the clients.

Now that I have approval for the general approach, I will proceed to finish this PR by plumbing this through the rest of the code base that I hadn't touched yet.

File: components/provider/src/lib.rs
Comment (sffc) [reply components/provider/src/lib.rs:115]: I will follow up here: https://github.com/unicode-org/icu4x/issues/422

File: components/provider/src/lib.rs
Comment (sffc) [reply components/provider/src/lib.rs:237]: I will follow up here: https://github.com/unicode-org/icu4x/issues/422

**Comment by sffc** [comment]
Okay, I migrated all ICU4X clients to the new DataProvider and deleted the old one.  I made a few additional changes that improve ergonomics.  I filed #422 to follow up on a generic version of DataProvider rather than a dyn trait object.  PTAL

**Review by zbraniecki — APPROVED** [review]

**Comment by sffc** [comment]
I'm going to check this in on Zibi's r+ to unblock my next PRs.  @Manishearth if you have further feedback, I'll do it in a follow-up PR.  Thanks!

**Pull request merged**
